Code/Tools/RemoteConsole/Core/RemoteConsoleCore.cpp

```diff
@@ -93,40 +93,63 @@ bool RCON_IsRemoteAllowedToConnect(const AZ::AzSock::AzSocketAddress& connectee)
 /////////////////////////////////////////////////////////////////////////////////////////////
 /////////////////////////////////////////////////////////////////////////////////////////////
 /////////////////////////////////////////////////////////////////////////////////////////////
+void SRemoteThreadedObject::Start(const char* name)
+{
+    AZStd::thread_desc desc;
+    desc.m_name = name;
+
+    auto function = AZStd::bind(&SRemoteThreadedObject::ThreadFunction, this);
+    m_thread = AZStd::thread(function, &desc);
+}
+
+void SRemoteThreadedObject::WaitForThread()
+{
+    if (m_thread.joinable())
+    {
+        m_thread.join();
+    }
+}
+
+void SRemoteThreadedObject::ThreadFunction()
+{
+    Run();
+    Terminate();
+}
 
+/////////////////////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////////////////////
 void SRemoteServer::StartServer()
 {
     StopServer();
     m_bAcceptClients = true;
-    Start(0, kServerThreadName);
+    Start(kServerThreadName);
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////////
 void SRemoteServer::StopServer()
 {
-    Stop();
     m_bAcceptClients = false;
     AZ::AzSock::CloseSocket(m_socket);
     m_socket = SOCKET_ERROR;
-    m_lock.Lock();
-    for (TClients::iterator it = m_clients.begin(); it != m_clients.end(); ++it)
     {
-        it->pClient->StopClient();
+        AZStd::lock_guard<AZStd::recursive_mutex> lock(m_mutex);
+        for (TClients::iterator it = m_clients.begin(); it != m_clients.end(); ++it)
+        {
+            it->pClient->StopClient();
+        }
     }
-    m_lock.Unlock();
-    m_stopEvent.Wait();
-    m_stopEvent.Set();
-}
+    AZStd::unique_lock<AZStd::recursive_mutex> lock(m_mutex);
+    m_stopCondition.wait(lock, [this] { return m_clients.empty(); });}
```

Comment:
The lock shouldn't be relased and then re-acquired here.

It should be maintained until the call to wait on the condition variable

```
    AZStd::scoped_lock lock(m_mutex);
    for (TClients::iterator it = m_clients.begin(); it != m_clients.end(); ++it)
    {
        it->pClient->StopClient();
    }
    m_stopCondition.wait(lock, [this] { return m_clients.empty(); });}
```

Reply:
`AZStd::scoped_lock` doesn't provide the necessary mechanism to manually control the lock inside the condition variable though.  Will instead move the `AZStd::unique_lock` up to capture both parts.

Reply:
That is my mistake, [condition_variable::wait](https://github.com/o3de/o3de/commit/ecd3b015eea1eb49bab8d0eb02f43b16054f82c4) only accepts a unique_lock

Reply:
On second thought, the lock wrapper can be removed entirely and mutex used directly since it's a `condition_variable_any`

Reply:
We should still use the lock wrapper, as the mutex will need to be released after the `m_stopCondition` has been notified.